File: xamarin/android/PropertyFinder/Utils/DownloadImageTask.cs
using System;
using Android.OS;
using Android.Widget;
using Android.Graphics;
using System.Net;
using System.IO;

namespace PropertyFinder
{
	public class DownloadImageTask : AsyncTask
	{
		private WeakReference imgRef;

		public String Url
		{
			get;
			private set;
		}

		public DownloadImageTask(ImageView image)
		{
			imgRef = new WeakReference(image);
		}

		protected override Java.Lang.Object DoInBackground(params Java.Lang.Object[] args)
		{
			Url = args[0].ToString();
			Bitmap bitmap = null;
			try
			{
				HttpWebRequest request = (HttpWebRequest) WebRequest.Create(Url);
				HttpWebResponse response = (HttpWebResponse) request.GetResponse();

				using(Stream s = response.GetResponseStream())
				{
					bitmap = BitmapFactory.DecodeStream(s);
				}
			}
			catch(Exception e)
			{
				// log
			}
			return bitmap;
		}

		protected override void OnPostExecute(Java.Lang.Object result)
		{
			Bitmap bitmap = result as Bitmap;

			if (IsCancelled)
			{
				bitmap = null;
			}
			if (imgRef != null && bitmap != null)
			{
				ImageView img = (ImageView) imgRef.Target;
				DownloadImageTask task = BitmapUtils.GetTask(img);

				if(this == task && img != null)
				{
					img.SetImageBitmap(bitmap);
				}
			}
		}
	}
}

File: native/android/PropertyCross/src/com/propertycross/android/util/DownloadImageTask.java
package com.propertycross.android.util;

import java.io.BufferedInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.lang.ref.WeakReference;
import java.net.HttpURLConnection;
import java.net.URL;

import uk.co.senab.bitmapcache.BitmapLruCache;
import uk.co.senab.bitmapcache.CacheableBitmapDrawable;

import android.graphics.BitmapFactory;
import android.os.AsyncTask;
import android.util.Log;
import android.widget.ImageView;

public class DownloadImageTask extends AsyncTask<String, Void, CacheableBitmapDrawable> {

    private static final String TAG = "DownloadImageTask";
    private final BitmapLruCache cache;
    private final WeakReference<ImageView> imageRef;
    private final BitmapFactory.Options decodeOptions;

    DownloadImageTask(ImageView imageView, BitmapLruCache cache, BitmapFactory.Options decodeOpts) {
        this.cache = cache;
        imageRef = new WeakReference<ImageView>(imageView);
        decodeOptions = decodeOpts;
    }

    @Override
    protected CacheableBitmapDrawable doInBackground(String... params) {
        try {
            if (imageRef.get() == null) {
                return null;
            }

            final String url = params[0];
            CacheableBitmapDrawable result = cache.get(url, decodeOptions);

            if (result == null) {
                Log.d(TAG, "Downloading image from: " + url);

                HttpURLConnection conn = (HttpURLConnection) new URL(url).openConnection();
                InputStream is = new BufferedInputStream(conn.getInputStream());
                result = cache.put(url, is, decodeOptions);
            } else {
                Log.d(TAG, "Got from Cache: " + url);
            }

            return result;

        } catch (IOException e) {
            Log.e(TAG, e.toString());
        }

        return null;
    }

    @Override
    protected void onPostExecute(CacheableBitmapDrawable result) {
        super.onPostExecute(result);

        ImageView iv = imageRef.get();
        if (iv != null) {
            iv.setImageDrawable(result);
        }
    }
}
